### PR TITLE
dsm manual checkpoint tests

### PR DESCRIPTION
**Pull request by wconti27**
## Workflow

1. ⚠️ Create your PR as draft ⚠️
2. Work on you PR until the CI passes (if something not related to your task is failing, you can ignore it)
3. Mark it as ready for review
    * Test logic is modified? -> Get a review from RFC owner. We're working on refining the `codeowners` file quickly.
    * Framework is modified, or non obvious usage of it -> get a review from [R&P team](https://dd.enterprise.slack.com/archives/C025TJ4RZ8X)

:rocket: Once your PR is reviewed, you can merge it!

🛟 [#apm-shared-testing](https://dd.enterprise.slack.com/archives/C025TJ4RZ8X) 🛟

## Reviewer checklist

* [ ] If PR title starts with `[<language>]`, double-check that only `<language>` is impacted by the change
* [ ] No system-tests internal is modified. Otherwise, I have the approval from [R&P team](https://dd.enterprise.slack.com/archives/C025TJ4RZ8X)
* [ ] CI is green, or failing jobs are not related to this change (and you are 100% sure about this statement)
* [ ] A docker base image is modified?
    * [ ] the relevant `build-XXX-image` label is present
* [ ] A scenario is added (or removed)?
    * [ ] Get a review from [R&P team](https://dd.enterprise.slack.com/archives/C025TJ4RZ8X)

### CODE REVIEW

**Review by piochelepiotr — APPROVED** [review]

File: tests/integrations/test_dsm.py [hunk 1]
```diff
@@ -432,6 +432,149 @@ def test_dsmcontext_extraction_base64(self):
...
+    def setup_dsm_manual_checkpoint_intra_process(self):
+        self.produce = weblog.get(
+            f"/dsm/manual/produce?type=dd-streams&target=system-tests-queue", timeout=DSM_REQUEST_TIMEOUT,
+        )
+        self.consume = weblog.get(
+            f"/dsm/manual/consume?type=dd-streams&source=system-tests-queue&headers={self.produce.text}",
+            timeout=DSM_REQUEST_TIMEOUT,
+        )
+
+    def test_dsm_manual_checkpoint_intra_process(self):
```
Comment (tabgok) [inline tests/integrations/test_dsm.py:449]: I believe what we want to test here is checkpointing *without* going through a queue

Comment (wconti27) [reply tests/integrations/test_dsm.py:449]: if you look at the implemented endpoints, there is no "queue" the message is going through, its just setting a manual consume/produce checkpoint.

File: utils/_features.py [hunk 2]
```diff
@@ -2277,6 +2277,16 @@ def container_guardrail(test_object):
         pytest.mark.features(feature_id=309)(test_object)
         return test_object
 
+    @staticmethod
+    def datastreams_monitoring_support_for_manual_checkpoints(test_object):
+        """
+        Ensure DSM Manual Checkpointing API is satisfied
+
+        https://feature-parity.us1.prod.dog/#/?feature=310
```
[the remark this answers is not in the document]
Comment (wconti27) [reply utils/_features.py]: Fixed

File: utils/build/docker/java/spring-boot/src/main/java/com/datadoghq/system_tests/springboot/App.java [hunk 3]
```diff
@@ -654,6 +660,106 @@ String dsmExtract(
         return "ok";
     }
 
+    @RequestMapping("/dsm/manual/produce")
```
[the remark this answers is not in the document]
Comment (wconti27) [reply utils/build/docker/java/spring-boot/src/main/java/com/datadoghq/system_tests/springboot/App.java:663]: Aye Aye Sir 🫡

**Review by cbeauchesne — CHANGES_REQUESTED** [review]

File: manifests/java.yml [hunk 4]
```diff
@@ -1158,6 +1158,12 @@ tests/:
       Test_DsmSQS:
         "*": irrelevant
         spring-boot: v0.1 # real version not known
+      Test_Dsm_Manual_Checkpoint_Inter_Process:
+        "*": irrelevant
+        spring-boot: bug (DSM for Java flips between using local empty DSM context and extracted context)
```
[the remark this answers is not in the document]
Comment (wconti27) [reply manifests/java.yml:1163]: done, here is the ticket I've created: https://datadoghq.atlassian.net/browse/AIDM-325

**Review by cbeauchesne — CHANGES_REQUESTED** [review]

**Review by cbeauchesne — APPROVED** [review]

**Review by tabgok — APPROVED** [review]

File: utils/build/docker/java/spring-boot/src/main/java/com/datadoghq/system_tests/springboot/App.java [hunk 5]
```diff
@@ -654,6 +660,111 @@ String dsmExtract(
...
+    String dsmManualCheckpointProduce(
```
Comment (amarziali) [inline utils/build/docker/java/spring-boot/src/main/java/com/datadoghq/system_tests/springboot/App.java:664]: Nit: you can avoid having `HttpServletResponse` as a parameter and just use `ResponseEntity<String>` as return type (https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/http/ResponseEntity.html)

File: utils/build/docker/java/spring-boot/src/main/java/com/datadoghq/system_tests/springboot/App.java [hunk 6]
```diff
@@ -654,6 +660,111 @@ String dsmExtract(
...
+
+        // Add headers that include DSM pathway context to response headers
+        for (Map.Entry<String, Object> entry : headers.entries()) {
+            response.addHeader(entry.getKey(), entry.getValue().toString());
+        }
+        return "ok";
+    }
+
+    @RequestMapping("/dsm/manual/produce_with_thread")
+    String dsmManualCheckpointProduceWithThread(
```
Comment (amarziali) [inline utils/build/docker/java/spring-boot/src/main/java/com/datadoghq/system_tests/springboot/App.java:685]: what's the aim of this test/endpoint? The context propagation through an executor service is already broadly tested in the java tracer

File: utils/build/docker/java/spring-boot/src/main/java/com/datadoghq/system_tests/springboot/data_streams/DSMContextCarrier.java [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    public DSMContextCarrier() {
+      this.data = new ConcurrentHashMap<>();
+    }
+
+    public DSMContextCarrier(Map<String, Object> data) {
+      this.data = data;
+    }
+
+    @Override
+    public synchronized Set<Map.Entry<String, Object>> entries() {
```
Comment (amarziali) [inline utils/build/docker/java/spring-boot/src/main/java/com/datadoghq/system_tests/springboot/data_streams/DSMContextCarrier.java:23]: there is already a concurrent hashmap 
Perhaps having a synchronized method is not needed here

**Review by amarziali — APPROVED** [review]
looks ok. I left some suggestions to improve the java apps under test

**Review by nayeem-kamal — APPROVED** [review]

File: utils/build/docker/java/spring-boot/src/main/java/com/datadoghq/system_tests/springboot/App.java (hunk 6 above)
Comment (wconti27) [reply utils/build/docker/java/spring-boot/src/main/java/com/datadoghq/system_tests/springboot/App.java:685]: to ensure the api for data streams manual checkpointing is consistent across langs (in this case, that is handles multi-process code)

File: utils/build/docker/java/spring-boot/src/main/java/com/datadoghq/system_tests/springboot/App.java (hunk 5 above)
Comment (wconti27) [reply utils/build/docker/java/spring-boot/src/main/java/com/datadoghq/system_tests/springboot/App.java:664]: Thanks!

**Pull request merged**